File: python/영어끝말잇기.py
def solution(n, words):
    answer = []
    wrong = False

    stack = []
    for i in words:
        if len(stack) < 1:
            stack.append(i)
            continue
        if i in stack or stack[len(stack)-1][len(stack[len(stack)-1])-1] != i[0]:
            stack.append(i)
            wrong = True
            break
        stack.append(i)
    if not wrong:
        return [0, 0]
    answer.append(len(stack)%n)
    if answer == [0]:
        answer[0] = n
    if len(stack)%n == 0:
        answer.append(len(stack)//n)
    else:
        answer.append(len(stack)//n+1)
    return answer

print(solution(3, ["tank", "kick", "know", "wheel", "land", "dream", "mother", "robot", "tank"]))
print(solution(5, ["hello", "observe", "effect", "take", "either", "recognize", "encourage", "ensure", "establish", "hang", "gather", "refer", "reference", "estimate", "executive"]))
print(solution(2, ["hello", "one", "even", "never", "now", "world", "draw"]))
print(solution(2, ["hello", "two"]))

File: JavaScript/영어끝말잇기.js
function solution(n, words) {
    let speaker = 1;
    let count = 0;
    const list = [];
    for (let i = 0; i < words.length; i++) {
        count += 1;
        let pWord = i > 0 ? words[i - 1] : null;
        const cWord = words[i];
        if (list.includes(cWord) || (pWord && pWord[pWord.length - 1] !== cWord[0])) {
            return [speaker, Math.ceil(count / n)];
        }
        list.push(words[i]);
        speaker = speaker === n ? 1 : speaker + 1;
    }
    return [0, 0];
}

//console.log(solution(3, ["tank", "kick", "know", "wheel", "land", "dream", "mother", "robot", "tank"]));
console.log(solution(5, ["hello", "observe", "effect", "take", "either", "recognize", "encourage", "ensure", "establish", "hang", "gather", "refer", "reference", "estimate", "executive"]));
//console.log(solution(2, ["hello", "one", "even", "never", "now", "world", "draw"]));
